scripts/ci/upload-r2-object.py

```
@@ -0,0 +1,138 @@
+
+
+def _build_signed_request(args: argparse.Namespace, body: bytes, amz_date: str) -> urllib.request.Request:
+    access_key = os.environ.get("AWS_ACCESS_KEY_ID", "")
+    secret_key = os.environ.get("AWS_SECRET_ACCESS_KEY", "")
+    region = os.environ.get("AWS_DEFAULT_REGION", "auto")
+    if not access_key or not secret_key:
+        raise SystemExit("AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are required")
+
+    parsed = urllib.parse.urlsplit(args.endpoint_url.rstrip("/"))
+    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
+        raise SystemExit(f"Invalid R2 endpoint URL: {args.endpoint_url}")
```

Ruff (0.15.17)

[warning] Avoid specifying long messages outside the exception class

(TRY003)